### PR TITLE
Cleanup RotaryEmbedding implementation + Add SinusoidalEmbedding

### PR DESCRIPTION
**Pull request by vinamarora8**
Mainly: General cleanup of RotaryEmbedding implementation.

1. Creates a `torch_brain/nn/position_embeddings.py` that houses `SinusoidalEmbedding` and `RotaryEmbedding` (sinusoidal embeddings come for ~free with time period computation code, so why not).
2. Rotary embedding now includes both `t_min` and `t_max` in its computed periods. To account for this, the default `t_max` values of `POYO` and `POYOPlus` were modified to their true original values.
3. This Rotary embedding forward pass computes the cosine/sine so that it does not happen in every attention layer (repeated computation). `Rotary*Attention` and tests were edited to accommodate this change.
4. An explicit `rotate_dim` parameter is provided so that we're not "rotating only half the dimensions" always.
5. General code cleanups/edits in the way that made sense to me. Feedback welcome.

wandb report with a NLB run on this PR vs main branch: https://api.wandb.ai/links/torch-brain/abkffpa9
300 epochs, batch size 32

### CODE REVIEW

File: torch_brain/nn/position_embeddings.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        x = rearrange(x, "... (d r) -> ... d r", r=2)
+        x1, x2 = x.unbind(dim=-1)
+        x = torch.stack((-x2, x1), dim=-1)
+        return rearrange(x, "... d r -> ... (d r)")
+
+    @classmethod
+    def apply_rotary_emb(
+        cls,
+        rotary_emb: Tensor,
+        x: Tensor,
```
Comment (milosobral) [inline torch_brain/nn/position_embeddings.py:125]: [EXTEREMELY MINOR] I feel like in pytorch code, I usually expect the `x` (input data) to a module to come as a first argument. (this is nitpicky, feel free to ignore)

**Review by milosobral — COMMENTED** [review]
Looks like a great improvement for readability of that whole module!

File: torch_brain/nn/position_embeddings.py (hunk 1 above)
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py:125]: Hmmm, I never realized that; you're right. It's weird to see `x` not being the first arg. I'll update this.

**Review by mazabou — CHANGES_REQUESTED** [review]
I have left a few comments on my first pass. I will come back to this in two hours, to finish reviewing the changes to test and the SinusoidalEmbedding class

File: torch_brain/nn/position_embeddings.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            dimensions to be effective.
+        t_min (float): Minimum period of the sinusoids. Set this to the smallest
+            timescale the attention layer should care about.
+        t_max (float): Maximum period of the sinusoids. Set this to the largest
+            timescale the attention layer should care about.
+    """
+
+    omega: Tensor
+
+    def __init__(self, head_dim: int, rotate_dim: int, t_min: float, t_max: float):
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py:86]: It was perceiverIO that did the half-dimension rotation, not the OG rotary paper.

File: torch_brain/nn/position_embeddings.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+    @staticmethod
+    def _rotate_half(x: Tensor) -> Tensor:
+        x = rearrange(x, "... (d r) -> ... d r", r=2)
+        x1, x2 = x.unbind(dim=-1)
+        x = torch.stack((-x2, x1), dim=-1)
+        return rearrange(x, "... d r -> ... (d r)")
+
+    @classmethod
+    def apply_rotary_emb(
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py:122]: I like `.rotate`. I'll change it to that

File: torch_brain/nn/position_embeddings.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py:86]: For SinusoidalEmbeddings, actually any `dim` is allowed since the user can add or concat these to the tokens. 
For RotaryEmbedding, this _has_ to be the attention head dimension; which is why chose the name. Does this make sense to you?

File: torch_brain/nn/position_embeddings.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    omega: Tensor
+
+    def __init__(self, head_dim: int, rotate_dim: int, t_min: float, t_max: float):
+        super().__init__()
+
+        if rotate_dim % 2 != 0:
+            raise ValueError("rotate_dim must be a multiple of 2")
+
+        if not head_dim >= rotate_dim:
+            raise ValueError("head_dim must be equal to or larger than rotate_dim")
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py:97]: done

File: torch_brain/nn/position_embeddings.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+        periods = SinusoidalEmbedding.get_periods(rotate_dim // 2, t_min, t_max)
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py]: Good point. done!

File: torch_brain/nn/position_embeddings.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py:74]: it's part of the `get_timeperiods()` function

File: torch_brain/nn/position_embeddings.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        omega = torch.zeros(head_dim // 2)
+        omega[: rotate_dim // 2] = 2 * torch.pi / periods
+        self.register_buffer("omega", omega)
+
+    @torch.no_grad
+    @torch.autocast(device_type="cuda", enabled=False)
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py:82]: worth checking

File: torch_brain/nn/position_embeddings.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    def forward(self, timestamps: Tensor) -> Tensor:
+        r"""Computes the rotary embeddings for given timestamps,
+        which can then be used by :meth:`RotaryEmbedding.apply_rotary_emb`.
+
+        Args:
+            timestamps (torch.Tensor): timestamps tensor.
+        """
+        angles = timestamps.unsqueeze(-1) * self.omega
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py:90]: good point. ill make this edit

File: torch_brain/nn/position_embeddings.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        angles = repeat(angles, "... n -> ... (n r)", r=2)
+        pos_emb = torch.cat((angles.cos(), angles.sin()), dim=-1)
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py:92]: i'll change this to `rotary_emb`

File: torch_brain/nn/position_embeddings.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
+from typing import Union
+import torch
+from torch import nn, Tensor
+from einops import repeat, rearrange
+
+
+class SinusoidalEmbedding(nn.Module):
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py:7]: ill rename these to *TimeEmbedding

File: torch_brain/nn/position_embeddings.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py]: done!

File: torch_brain/nn/position_embeddings.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        x1, x2 = x.unbind(dim=-1)
+        x = torch.stack((-x2, x1), dim=-1)
+        return rearrange(x, "... d r -> ... (d r)")
+
+    @classmethod
+    def apply_rotary_emb(
+        cls,
+        x: Tensor,
+        rotary_emb: Tensor,
+        head_dim: int = 2,
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py:106]: Talked about it on call as well. It can't be `rotation_dim`. We can think about what to name it for a bit

File: torch_brain/nn/position_embeddings.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py]: done!

File: torch_brain/nn/position_embeddings.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+    omega: Tensor
+
+    def __init__(self, dim: int, t_min: float, t_max: float):
+        super().__init__()
+        if dim % 2 != 0:
+            raise ValueError("`dim` must be a multiple of 2")
+
+        periods = self.get_periods(dim // 2, t_min, t_max)
+        omega = 2 * torch.pi / periods
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py:30]: we talked about it on call and decided to stay with timeperiods

File: torch_brain/nn/position_embeddings.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py:67]: We talked on call and decided to keep it this way.

File: torch_brain/nn/position_embeddings.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/nn/position_embeddings.py:82]: Lmk your final thoughts

**Comment by vinamarora8** [comment]
@mazabou I have confirmed that the poyo_mp checkpoint ([link](https://torch-brain.s3.amazonaws.com/model-zoo/poyo_mp.ckpt)) works perfectly
- It overwrites the default `omega` to whatever is in the checkpoint
- The results on this PR and the main branch are the same. Average val metric (before any training): 0.933331 for both versions of the code.

**Review by mazabou — CHANGES_REQUESTED** [review]
The changes look good thank you. 

I left a small suggestion + we still need to figure out how to name `head_dim`

**Review by mazabou — APPROVED** [review]
LGTM
